crates/node/src/sync.rs

```diff
@@ -23,25 +25,47 @@ pub struct SyncConfig {
     pub interval: Duration,
 }
 
-async fn send(stream: &mut Stream, message: &StreamMessage<'_>) -> EyreResult<()> {
-    let message = borsh::to_vec(message)?;
-
-    stream.send(Message::new(message)).await?;
+async fn send(
+    stream: &mut Stream,
+    message: &StreamMessage<'_>,
+    shared_key: Option<SharedKey>,
+) -> EyreResult<()> {
+    let base_data = borsh::to_vec(message)?;
+
+    let data = match shared_key {
+        Some(key) => match key.encrypt(base_data, [0; aead::NONCE_LEN]) {
+            Some(data) => data,
+            None => bail!("Encryption failed"),
+        },
+        None => base_data,
+    };
 
+    stream.send(Message::new(data)).await?;
     Ok(())
 }
 
 async fn recv(
     stream: &mut Stream,
     duration: Duration,
+    shared_key: Option<SharedKey>,
 ) -> EyreResult<Option<StreamMessage<'static>>> {
     let Some(message) = timeout(duration, stream.next()).await? else {
         return Ok(None);
     };
 
-    let message = borsh::from_slice(&message?.data)?;
+    let message_data = message?.data.to_vec();
```

Comment:
`to_vec` will copy to a new allocation. whereas messages received are already owned.

```suggestion
    let message_data = message?.data.into_owned();
```